src/bms/parse/validity.rs

```diff
@@ -0,0 +1,614 @@
+//! Validity checks for BMS data after parsing and/or manual edits.
+//!
+//! This module provides a set of structural validations that are independent of
+//! the parsing process. It can be used after editing `Bms` in-memory to ensure
+//! referential integrity and basic invariants required for correct playback.
+
+use thiserror::Error;
+
+use crate::bms::command::{
+    ObjId,
+    channel::{Key, NoteKind, PlayerSide},
+    time::ObjTime,
+};
+
+use super::model::{Bms, obj::Obj};
+
+/// Empty-related validity entries.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Error)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ValidityEmpty {
+    /// Text content is empty at the time.
+    #[error("Empty text content at {0:?}")]
+    EmptyTextAt(ObjTime),
+    /// Option content is empty at the time.
+    #[cfg(feature = "minor-command")]
+    #[error("Empty option content at {0:?}")]
+    EmptyOptionAt(ObjTime),
+}
+
+/// Missing-related validity entries.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Error)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum ValidityMissing {
```

Comment:
There are other missing cases to consider (even only major commands) more:

- Missing BPM change `#BPMxx`
- Missing scroll stop `#STOPxx`